src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -75,138 +74,236 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
-                        break;
+                        return NativeLibrary.TryLoad(hostfxrAssembly, out var handle) ? handle : IntPtr.Zero;
                     }
                 }
             }
 
-            if (dotnetPath is null)
-            {
-                throw new InvalidOperationException("Could not find the dotnet executable. Is it on the PATH?");
-            }
+            return IntPtr.Zero;
+        }
 
-            string bestSDK = null;
-            int rc = NativeMethods.hostfxr_resolve_sdk2(exe_dir: dotnetPath, working_dir: workingDirectory, flags: 0, result: (key, value) =>
+        private static string SdkResolutionExceptionMessage(string methodName) => $"Failed to find all versions of .NET Core MSBuild. Call to {methodName}. There may be more details in stderr.";
+        
+        /// <summary>
+        /// Determines the directory location of the SDK accounting for
+        /// global.json and multi-level lookup policy.
+        /// </summary>
+        private static string? GetSdkFromGlobalSettings(string workingDirectory)
+        {
+            string? resolvedSdk = null;
+            int rc = NativeMethods.hostfxr_resolve_sdk2(exe_dir: DotnetPath.Value, working_dir: workingDirectory, flags: 0, result: (key, value) =>
             {
                 if (key == NativeMethods.hostfxr_resolve_sdk2_result_key_t.resolved_sdk_dir)
                 {
-                    bestSDK = value;
+                    resolvedSdk = value;
                 }
             });
 
-            if (rc == 0 && bestSDK != null)
+            if (rc != 0)
             {
-                yield return bestSDK;
+                throw new InvalidOperationException(SdkResolutionExceptionMessage(nameof(NativeMethods.hostfxr_resolve_sdk2)));
             }
-            else if (rc != 0)
+
+            return resolvedSdk;
+        }
+
+        private static string ResolveDotnetPath()
+        {
+            string? dotnetPath = GetDotnetPathFromROOT();
+
+            if (string.IsNullOrEmpty(dotnetPath))
             {
-                throw new InvalidOperationException("Failed to find an appropriate version of .NET Core MSBuild. Call to hostfxr_resolve_sdk2 failed. There may be more details in stderr.");
+                string? dotnetExePath = GetCurrentProcessPath();
+                var isRunFromDotnetExecutable = !string.IsNullOrEmpty(dotnetExePath) 
+                    && Path.GetFileName(dotnetExePath).Equals(ExeName, StringComparison.InvariantCultureIgnoreCase);
+                
+                if (isRunFromDotnetExecutable)
+                {
+                    dotnetPath = Path.GetDirectoryName(dotnetExePath);
+                }
+                else
+                {
+                    dotnetPath = FindDotnetPathFromEnvVariable("DOTNET_HOST_PATH") 
+                        ?? FindDotnetPathFromEnvVariable("DOTNET_MSBUILD_SDK_RESOLVER_CLI_DIR")
+                        ?? GetDotnetPathFromPATH();
```

Comment:
We should have one last place where we just look in the expected places, right? Like /usr/bin/dotnet or whatever on linux?

Reply:
Chet listed the lookup sequence - based on the execution context of MsBuild.Locator, we check different places. More details are in the PR description.